hazelcast/src/main/java/com/hazelcast/cache/impl/HazelcastServerCachingProvider.java

```diff
@@ -85,30 +85,35 @@ protected HazelcastInstance getOrCreateInstance(URI uri, ClassLoader classLoader
             return HazelcastInstanceFactory.getOrCreateHazelcastInstance(config);
         }
 
-        // If instance name is specified via properties, get instance through it.
+        // If instance name is specified via properties, get instance through it or create a new instance with default config
+        // and given instance name
         if (instanceName != null) {
-            return Hazelcast.getHazelcastInstanceByName(instanceName);
+            HazelcastInstance instance = getOrCreateByInstanceName(instanceName);
+            return instance;
         }
 
         // resolving HazelcastInstance via properties failed, try with URI as XML configuration file location
         final boolean isDefaultURI = (uri == null || uri.equals(getDefaultURI()));
         if (!isDefaultURI) {
-            try {
-                // try locating a Hazelcast config at CacheManager URI
-                Config config = getConfigFromLocation(uri, classLoader, null);
-                return HazelcastInstanceFactory.getOrCreateHazelcastInstance(config);
-            } catch (Exception e) {
-                if (LOGGER.isFinestEnabled()) {
-                    LOGGER.finest("Could not get or create hazelcast instance from URI " + uri.toString(), e);
+            // attempt to resolve URI as config location or as instance name
+            if (isConfigLocation(uri)) {
```

Comment:
Just a thought - Looks like it is doing the right thing but I am thinking if we should add an early return and break down this method into a couple of helper methods for easier readability. As this is a backport I wouldn't bother with this here, perhaps in a separate PR or think about it in the future.
Something like this?
```
@Override
    protected HazelcastInstance getOrCreateInstance(URI uri, ClassLoader classLoader, Properties properties)
            throws URISyntaxException, IOException {
        HazelcastInstance instance = getOrCreateFromProperties(classLoader, properties);
        if (instance != null) {
            return instance;
        }

        // resolving HazelcastInstance via properties failed, try with URI as XML configuration file location
        if (uri == null || uri.equals(getDefaultURI())) {
            return getDefaultInstance();
        }

        // attempt to resolve URI as config location or as instance name
        final boolean isConfigLocation = isConfigLocation(uri);
        try {
            return isConfigLocation
                    ? HazelcastInstanceFactory.getOrCreateHazelcastInstance(getConfigFromLocation(uri, classLoader, null))
                    : getOrCreateByInstanceName(uri.toString());
        } catch (Exception e) {
            if (LOGGER.isFinestEnabled()) {
                final String msg = isConfigLocation
                        ? "Could not get or create hazelcast instance from URI " + uri.toString()
                        : "Could not get hazelcast instance from instance name" + uri.toString();
                LOGGER.finest(msg, e);
            }
        }
        // could not locate hazelcast instance, return null and an exception will be thrown from invoker
        return null;
    }

    private HazelcastInstance getOrCreateFromProperties(ClassLoader classLoader, Properties properties) throws URISyntaxException, IOException {
        HazelcastInstance instanceItself = (HazelcastInstance) properties.get(HAZELCAST_INSTANCE_ITSELF);

        // If instance itself is specified via properties, get instance through it.
        if (instanceItself != null) {
            return instanceItself;
        }

        String location = properties.getProperty(HAZELCAST_CONFIG_LOCATION);
        String instanceName = properties.getProperty(HAZELCAST_INSTANCE_NAME);

        // If config location is specified via properties, get instance through it.
        if (location != null) {
            Config config = getConfigFromLocation(location, classLoader, instanceName);
            return HazelcastInstanceFactory.getOrCreateHazelcastInstance(config);
        }

        // If instance name is specified via properties, get instance through it or create a new instance with default config
        // and given instance name
        if (instanceName != null) {
            return getOrCreateByInstanceName(instanceName);
        }
        return null;
    }
```